src/sbin/mount_vnd/mount_vnd.8
.\"	$OpenBSD: mount_vnd.8,v 1.13 2008/05/26 21:14:46 jmc Exp $
.\"
.\" Copyright (c) 1993 University of Utah.
.\" Copyright (c) 1980, 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Systems Programming Group of the University of Utah Computer
.\" Science Department.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)vnconfig.8	8.1 (Berkeley) 6/5/93
.\"
.\"
.\" Copyright (c) 2007 Alexander von Gernler <grunk@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: June 12 2008 $
.Dt MOUNT_VND 8
.Os
.Sh NAME
.Nm mount_vnd ,
.Nm vnconfig
.Nd configure vnode disks
.Sh SYNOPSIS
.Nm mount_vnd
.Bk -words
.Op Fl k
.Op Fl K Ar rounds
.Op Fl o Ar options
.Op Fl S Ar saltfile
.Ar image
.Ar vnd_dev
.Ek
.Nm vnconfig
.Bk -words
.Op Fl ckluv
.Op Fl K Ar rounds
.Op Fl S Ar saltfile
.Ar vnd_dev
.Ar image
.Ek
.Sh DESCRIPTION
The
.Nm vnconfig
command configures vnode pseudo disk devices.
It will associate (or disassociate) the special file
.Ar vnd_dev
with the regular file
.Ar image ,
allowing the latter to be accessed as though it were a disk.
Hence a regular file within the filesystem can be used for swapping
or can contain a filesystem that is mounted in the name space.
.Pp
Both traditional devices,
.Xr vnd 4 ,
and the cache-coherent devices,
.Pa svnd ,
are configured through
.Nm vnconfig .
.Pa svnd
should be mounted with the
.Xr mount 8
option
.Dq sync
if
.Ar image
is being read from,
to ensure consistency in the buffer cache.
For normal operation this is not necessary.
.Pp
.Nm mount_vnd
works similarly to
.Nm vnconfig ,
but it provides an interface that can be used by the
.Xr fstab 5
infrastructure, so that an
.Ar image
file can be configured to a device
.Ar node
while booting.
.Nm mount_vnd
defaults to the
.Pa svnd0
device.
.Pp
For
.Xr fstab 5
lines with type
.Dq ffs ,
the
.Dq noauto
option must be set to prevent a
.Xr mount 8
of the FFS partitions
before the necessary vnd devices are configured.
Also, the
.Dq fs_passno
field has to be set to 0 to prevent
.Xr fsck 8
from checking the filesystem for the same reasons.
.Pp
.Nm mount_vnd
is invoked by
.Xr mount 8
when using the following syntax:
.Bd -ragged -offset 4n
.Nm mount Op options
-t vnd
.Ar image Ar node
.Ed
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
.Nm vnconfig
only.
Configures the device.
If successful, references to
.Ar vnd_dev
will access the contents of
.Ar image .
This is the default operation.
.It Fl K Ar rounds
Associate an encryption key with the device.
All data will be encrypted using the Blowfish cipher before it is
written to the disk.
The user is asked for both a passphrase and the name of a salt file.
The salt file can also be specified on the command line using the
.Fl S
option.
The passphrase and salt are combined according to PKCS #5 PBKDF2 for the
specified number of
rounds to generate the actual key used.
.Ar rounds
is a number between 1000 and
.Dv INT_MAX .
DO NOT LOSE THE SALT FILE.
Encryption only works with
.Pa svnd .
.It Fl k
Associate an encryption key with the device.
All data will be encrypted using the Blowfish cipher before it is
written to the disk.
Encryption only works with
.Pa svnd .
.It Fl l
.Nm vnconfig
only.
List the (s)vnd devices and indicate which ones are in use.
If a specific
.Ar vnd_dev
is given, then only that one will be described.
.It Fl o Ar options
.Nm mount_vnd
only.
Options are specified with a
.Fl o
flag followed by a comma separated string of options.
See the
.Xr mount 8
man page for possible options and their meanings.
.Pp
At the moment,
.Fl o
is only here for compatibility reasons, but no use is made of supplied
options.
.It Fl S Ar saltfile
When
.Fl K
is used, specify the
.Pa saltfile .
If the salt filename is not specified using
.Fl S ,
it defaults to
.Ar image Ns .slt .
.It Fl u
.Nm vnconfig
only.
Unconfigures a
.Ar vnd_dev .
.It Fl v
.Nm vnconfig
only.
Print messages to stdout describing actions taken.
.El
.Sh FILES
.Bl -tag -width /etc/rvnd?? -compact
.It Pa /dev/{,r}{,s}vnd*
.El
.Sh EXAMPLES
Configure a CD-ROM or DVD image file as vnode disk svnd0
and mount the ISO 9660 filesystem contained in it:
.Bd -literal -offset indent
# vnconfig svnd0 /tmp/diskimage
# mount -t cd9660 /dev/svnd0c /mnt
.Ed
.Pp
Configure an encrypted image file as vnode disk svnd0 and mount the FFS
filesystem contained in the
.Sq a
partition of the disklabel.
The encryption key does not echo and must be entered every time
the vnode disk is configured.
.Bd -literal -offset indent
# vnconfig -k svnd0 /tmp/cryptimg
Encryption key:
# mount /dev/svnd0a /mnt
.Ed
.Pp
An equivalent
.Xr fstab 5
entry is:
.Bd -literal -offset indent
/tmp/cryptimg /dev/svnd0c vnd rw,noauto,-k       0 0
/dev/svnd0a   /mnt        ffs rw,noauto          0 0
.Ed
.Pp
Same as above, but now configure the vnode using PKCS #5 PBKDF2 and
a salt file with 20000 rounds:
.Bd -literal -offset indent
# vnconfig -K 20000 svnd0 /tmp/cryptimg
Encryption key:
Salt file: /tmp/cryptsalt
# mount /dev/svnd0a /mnt
.Ed
.Pp
An equivalent
.Xr fstab 5
entry is:
.Bd -literal -offset indent
/tmp/cryptimg /dev/svnd0c vnd rw,noauto,-K=20000 0 0
/dev/svnd0a   /mnt        ffs rw,noauto          0 0
.Ed
.Pp
Unmount the filesystem and unconfigure the vnode device:
.Bd -literal -offset indent
# umount /mnt
# vnconfig -u svnd0
.Ed
.Pp
A sample of commands to put in
.Xr rc.local 8
in order to have vnd images configured at boot time is listed here:
.Bd -literal -offset indent
# Example for automatically configuring a vnd device on startup
echo "configuring vnd devices:"
mount /dev/svnd0c
fsck -p /dev/svnd0a
mount /mnt
.Ed
.Pp
Mounting images during the first pass of
.Xr fsck 8
and
.Xr mount 8
is not possible, because the image to be configured to a vnd itself
resides on a filesystem that first has to be checked and mounted.
.Sh SEE ALSO
.Xr vnd 4 ,
.Xr fstab 5 ,
.Xr mount 8 ,
.Xr swapon 8 ,
.Xr umount 8
.Sh HISTORY
The
.Nm
command first appeared in
.Ox 4.2 .
.Sh AUTHORS
.An David Krause .
.Sh CAVEATS
The
.Sq c
partition of a vnd image should not be used.
When a superblock becomes damaged,
.Xr fsck_ffs 8
needs information contained in the disklabel
to determine the location of alternate superblocks.
This information is not available when directly using the
.Sq c
partition,
so checking the filesystem image will fail.


src/sbin/mount_vnd/mount_vnd.c
/*	$OpenBSD: mount_vnd.c,v 1.4 2007/09/29 07:10:09 fkr Exp $	*/
/*
 * Copyright (c) 1993 University of Utah.
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * the Systems Programming Group of the University of Utah Computer
 * Science Department.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * from: Utah $Hdr: vnconfig.c 1.1 93/12/15$
 *
 *	@(#)vnconfig.c	8.1 (Berkeley) 12/15/93
 */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/mount.h>
#include <sys/stat.h>

#include <dev/vndioctl.h>

#include <blf.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <util.h>

#include "pkcs5_pbkdf2.h"

__RCSID("$MirOS: src/sbin/mount_vnd/mount_vnd.c,v 1.2 2008/06/12 21:53:44 tg Exp $");
__RCSID("$OpenBSD: mount_vnd.c,v 1.4 2007/09/29 07:10:09 fkr Exp $");
__RCSID("Utah $Hdr: vnconfig.c 1.1 93/12/15$");
__SCCSID("@(#)vnconfig.c	8.1 (Berkeley) 12/15/93");

#define DEFAULT_VND	"svnd0"

#define VND_CONFIG	1
#define VND_UNCONFIG	2
#define VND_GET		3

int verbose = 0;
int run_mount_vnd = 0;

__dead void	 usage(void);
int		 config(char *, char *, int, char *, size_t);
int		 getinfo(const char *);
char		*get_pkcs_key(char *, char *);

int
main(int argc, char **argv)
{
	int	 ch, rv, action, opt_c, opt_k, opt_K, opt_l, opt_u;
	char	*key, *mntopts, *rounds, *saltopt;
	size_t	 keylen = 0;
	extern char *__progname;

	if (strcasecmp(__progname, "mount_vnd") == 0)
		run_mount_vnd = 1;

	opt_c = opt_k = opt_K = opt_l = opt_u = 0;
	key = mntopts = rounds = saltopt = NULL;
	action = VND_CONFIG;

	while ((ch = getopt(argc, argv, "ckK:lo:S:uv")) != -1) {
		switch (ch) {
		case 'c':
			opt_c = 1;
			break;
		case 'k':
			opt_k = 1;
			break;
		case 'K':
			opt_K = 1;
			rounds = optarg;
			break;
		case 'l':
			opt_l = 1;
			break;
		case 'o':
			mntopts = optarg;
			break;
		case 'S':
			saltopt = optarg;
			break;
		case 'u':
			opt_u = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (opt_c + opt_l + opt_u > 1)
		errx(1, "-c, -l and -u are mutually exclusive options");

	if (opt_l)
		action = VND_GET;
	else if (opt_u)
		action = VND_UNCONFIG;
	else
		action = VND_CONFIG;	/* default behavior */

	if (saltopt && (!opt_K))
		errx(1, "-S only makes sense when used with -K");

	if (action == VND_CONFIG && argc == 2) {
		int ind_raw, ind_reg;

		if (opt_k) {
			if (opt_K)
				errx(1, "-k and -K are mutually exclusive");
			key = getpass("Encryption key: ");
			if (key == NULL || (keylen = strlen(key)) == 0)
				errx(1, "Need an encryption key");
		} else if (opt_K) {
			key = get_pkcs_key(rounds, saltopt);
			keylen = BLF_MAXUTILIZED;
		}

		/* fix order of arguments. */
		if (run_mount_vnd) {
			ind_raw = 1;
			ind_reg = 0;
		} else {
			ind_raw = 0;
			ind_reg = 1;
		}
		rv = config(argv[ind_raw], argv[ind_reg], action, key, keylen);
	} else if (action == VND_UNCONFIG && argc == 1)
		rv = config(argv[0], NULL, action, NULL, 0);
	else if (action == VND_GET)
		rv = getinfo(argc ? argv[0] : NULL);
	else
		usage();

	exit(rv);
}

char *
get_pkcs_key(char *arg, char *saltopt)
{
	char		 keybuf[128], saltbuf[128], saltfilebuf[PATH_MAX];
	char		*saltfile;
	char		*key = NULL;
	const char	*errstr;
	int		 rounds;

	rounds = strtonum(arg, 1000, INT_MAX, &errstr);
	if (errstr)
		err(1, "rounds: %s", errstr);
	key = getpass("Encryption key: ");
	if (!key || strlen(key) == 0)
		errx(1, "Need an encryption key");
	strncpy(keybuf, key, sizeof(keybuf));
	if (saltopt)
		saltfile = saltopt;
	else {
		printf("Salt file: ");
		fflush(stdout);
		saltfile = fgets(saltfilebuf, sizeof(saltfilebuf), stdin);
		if (saltfile)
			saltfile[strcspn(saltfile, "\n")] = '\0';
	}
	if (!saltfile || saltfile[0] == '\0') {
		warnx("Skipping salt file, insecure");
		memset(saltbuf, 0, sizeof(saltbuf));
	} else {
		int fd;

		fd = open(saltfile, O_RDONLY);
		if (fd == -1) {
			int *s;

			fprintf(stderr, "Salt file not found, attempting to create one\n");
			fd = open(saltfile, O_RDWR|O_CREAT|O_EXCL, 0600);
			if (fd == -1)
				err(1, "Unable to create salt file: '%s'",
				    saltfile);
			arc4random_buf(saltbuf, sizeof(saltbuf));
			if (write(fd, saltbuf, sizeof(saltbuf))
			    != sizeof(saltbuf))
				err(1, "Unable to write salt file: '%s'", saltfile);
			fprintf(stderr, "Salt file created as '%s'\n", saltfile);
		} else {
			if (read(fd, saltbuf, sizeof(saltbuf))
			    != sizeof(saltbuf))
				err(1, "Unable to read salt file: '%s'", saltfile);
		}
		close(fd);
	}
	if (pkcs5_pbkdf2((u_int8_t**)&key, BLF_MAXUTILIZED, keybuf,
	    sizeof(keybuf), saltbuf, sizeof(saltbuf), rounds, 0))
		errx(1, "pkcs5_pbkdf2 failed");

	return (key);
}

int
getinfo(const char *vname)
{
	int vd, print_all = 0;
	struct vnd_user vnu;

	if (vname == NULL) {
		vname = DEFAULT_VND;
		print_all = 1;
	}

	vd = opendev((char *)vname, O_RDONLY, OPENDEV_PART, NULL);
	if (vd < 0)
		err(1, "open: %s", vname);

	vnu.vnu_unit = -1;

query:
	if (ioctl(vd, VNDIOCGET, &vnu) == -1) {
		if (print_all && errno == ENXIO && vnu.vnu_unit > 0) {
			close(vd);
			return (0);
		} else {
			err(1, "ioctl: %s", vname);
		}
	}

	fprintf(stdout, "vnd%d: ", vnu.vnu_unit);

	if (!vnu.vnu_ino)
		fprintf(stdout, "not in use\n");
	else
		fprintf(stdout, "covering %s on %s, inode %d\n", vnu.vnu_file,
		    devname(vnu.vnu_dev, S_IFBLK), vnu.vnu_ino);

	if (print_all) {
		vnu.vnu_unit++;
		goto query;
	}

	close(vd);

	return (0);
}

int
config(char *dev, char *file, int action, char *key, size_t keylen)
{
	struct vnd_ioctl vndio;
	FILE *f;
	char *rdev;
	int rv;

	if (opendev(dev, O_RDONLY, OPENDEV_PART, &rdev) < 0)
		err(4, "%s", rdev);
	f = fopen(rdev, "r");
	if (f == NULL) {
		warn("%s", rdev);
		rv = -1;
		goto out;
	}
	vndio.vnd_file = file;
	vndio.vnd_key = (u_char *)key;
	vndio.vnd_keylen = keylen;

	/*
	 * Clear (un-configure) the device
	 */
	if (action == VND_UNCONFIG) {
		rv = ioctl(fileno(f), VNDIOCCLR, &vndio);
		if (rv)
			warn("VNDIOCCLR");
		else if (verbose)
			printf("%s: cleared\n", dev);
	}
	/*
	 * Configure the device
	 */
	if (action == VND_CONFIG) {
		rv = ioctl(fileno(f), VNDIOCSET, &vndio);
		if (rv)
			warn("VNDIOCSET");
		else if (verbose)
			printf("%s: %llu bytes on %s\n", dev, vndio.vnd_size,
			    file);
	}

	fclose(f);
	fflush(stdout);
 out:
	if (key)
		memset(key, 0, keylen);
	return (rv < 0);
}

__dead void
usage(void)
{
	extern char *__progname;

	if (run_mount_vnd)
		(void)fprintf(stderr,
		    "usage: %s [-k] [-K rounds] [-o options] "
		    "[-S saltfile] image vnd_dev\n", __progname);
	else
		(void)fprintf(stderr,
		    "usage: %s [-ckluv] [-K rounds] [-S saltfile] vnd_dev "
		    "image\n", __progname);

	exit(1);
}
